components/core/src/clp/Grep.cpp

```diff
@@ -1063,4 +1011,447 @@ size_t Grep::search(Query const& query, size_t limit, Archive& archive, File& co
 
     return num_matches;
 }
+
+void Grep::generate_query_substring_logtypes(
+        string& processed_search_string,
+        ByteLexer& lexer,
+        vector<std::set<QueryLogtype>>& query_substr_logtypes
+) {
+    // We need to differentiate between literal '*'/'?' and wildcards
+    auto [is_greedy_wildcard, is_non_greedy_wildcard, is_escape]
+            = get_wildcard_and_escape_locations(processed_search_string);
+
+    // Consider each substr(begin_idx,end_idx) of the processed_search_string and determine if it
+    // could have been compressed as static-text, a variable, or some combination of
+    // variables/static-text Then we populate each entry in query_substr_logtypes which corresponds
+    // to the logtype for substr(0,n). To do this, for each combination of substr(begin_idx,end_idx)
+    // that reconstructs substr(0,n) (e.g., substring "*1 34", can be reconstructed from substrings
+    // "*1", " ", "34"), store all possible logtypes (e.g. "*<int> <int>, "*<has#> <int>, etc.) that
+    // are unique from any previously checked combination. Each entry in query_substr_logtypes is
+    // used to build the following entry, with the last entry having all possible logtypes for the
+    // full query itself.
+    for (size_t end_idx = 1; end_idx <= processed_search_string.size(); ++end_idx) {
+        // Skip strings that end with an escape character (e.g., substring " text\" from string
+        // "* text\* *").
+        if (is_escape[end_idx - 1]) {
+            continue;
+        }
+        for (size_t begin_idx = 0; begin_idx < end_idx; ++begin_idx) {
+            // Skip strings that begin with an incorrectly unescaped wildcard (e.g., substring
+            // "*text" from string "* \*text *").
+            if ((begin_idx > 0 && is_escape[begin_idx - 1])) {
+                continue;
+            }
+            std::vector<QueryLogtype> possible_substr_types;
```

Comment:
I think we can extract the computation of `possible_substr_types` into its own method so that this method is easier to read (less indentation, etc.).